packages/upstream/conduit-async.2.2.2-1/files/0001-backport-Add-openssl_overrides-to-conduit-context.patch

```diff
@@ -0,0 +1,202 @@
+ 
+-let init ?src ?(tls_own_key=`None) ?(tls_server_key=`None) () =
++let init ?src ?(tls_own_key=`None) ?(tls_server_key=`None) ?openssl_overrides () =
+   let tls_own_key =
+     match tls_own_key with `None -> tls_server_key | _ -> tls_own_key in
+   match src with
+   | None ->
+-    Lwt.return { src=None; tls_own_key }
++    Lwt.return { src=None; tls_own_key; openssl_overrides; }
+   | Some host ->
+     let open Unix in
+     Lwt_unix.getaddrinfo host "0" [AI_PASSIVE; AI_SOCKTYPE SOCK_STREAM]
+     >>= function
+-    | {ai_addr;_}::_ -> Lwt.return { src=Some ai_addr; tls_own_key }
++    | {ai_addr;_}::_ -> Lwt.return { src=Some ai_addr; tls_own_key; openssl_overrides; }
+     | [] -> Lwt.fail_with "Invalid conduit source address specified"
+ 
+ module Sockaddr_io = struct
+@@ -270,6 +271,18 @@ let connect_with_openssl ~ctx (`Hostname hostname, `IP ip, `Port port) =
+         in
+         Some ctx_ssl
+   in
++  let hostname, ctx_ssl =
++    match ctx.openssl_overrides with
++    | None | Some { client = None } -> (hostname, ctx_ssl)
++    | Some { client = Some overrides } ->
++        let hostname =
++          match overrides.hostname with Some x -> x | None -> hostname
++        in
++        let ctx_ssl =
++          match overrides.ctx with Some x -> Some x | None -> ctx_ssl
++        in
++        (hostname, ctx_ssl)
++  in
+   Conduit_lwt_unix_ssl.Client.connect ?ctx:ctx_ssl ?src:ctx.src ~hostname sa
+   >>= fun (fd, ic, oc) ->
+   let flow = TCP {fd;ip;port} in
+diff --git a/lwt-unix/conduit_lwt_unix.mli b/lwt-unix/conduit_lwt_unix.mli
+index e5db5d0..37ad751 100644
+--- a/lwt-unix/conduit_lwt_unix.mli
++++ b/lwt-unix/conduit_lwt_unix.mli
+@@ -166,6 +166,7 @@ val init :
+   ?src:string ->
+   ?tls_own_key:tls_own_key ->
+   ?tls_server_key:tls_own_key (* Deprecated, use tls_own_key. *) ->
++  ?openssl_overrides:Conduit_lwt_unix_ssl.Overrides.t ->
+   unit -> ctx io
+ 
+ (** [connect ~ctx client] establishes an outgoing connection
+diff --git a/lwt-unix/conduit_lwt_unix_ssl_dummy.ml b/lwt-unix/conduit_lwt_unix_ssl_dummy.ml
+index b11f8d2..6e7d57d 100644
+--- a/lwt-unix/conduit_lwt_unix_ssl_dummy.ml
++++ b/lwt-unix/conduit_lwt_unix_ssl_dummy.ml
+@@ -15,6 +15,14 @@
+  *
+  *)
+ 
++module Overrides = struct
++  module Client = struct
++    type t = { ctx : [ `Ssl_not_available ] option ; hostname : string option }
++  end
++
++  type t = { client : Client.t option }
++end
++
+ module Client = struct
+   let default_ctx = `Ssl_not_available
+ 
+diff --git a/lwt-unix/conduit_lwt_unix_ssl_dummy.mli b/lwt-unix/conduit_lwt_unix_ssl_dummy.mli
+index 24b73b0..cf25544 100644
+--- a/lwt-unix/conduit_lwt_unix_ssl_dummy.mli
++++ b/lwt-unix/conduit_lwt_unix_ssl_dummy.mli
+@@ -17,6 +17,14 @@
+ 
+ (** TLS/SSL connections via {{:http://www.openssl.org}OpenSSL} C bindings *)
+ 
++module Overrides : sig
++  module Client : sig
++    type t = { ctx : [ `Ssl_not_available ] option ; hostname : string option }
```

Comment:
This is using polymorphic variants. Is it not possible regular variants which create better error messages?

Reply:
I believe it needs to be this otherwise it won't compile (the library uses polymorphic variants everywhere)